src/v/datalake/coordinator/state_update.cc

```diff
@@ -158,4 +227,77 @@ mark_files_committed_update::apply(topics_state& state) {
     return std::nullopt;
 }
 
+checked<bool, stm_update_error>
+topic_lifecycle_update::can_apply(const topics_state& state) {
+    auto topic_it = state.topic_to_state.find(topic);
+    if (topic_it == state.topic_to_state.end()) {
+        return true;
+    }
+    auto& cur_state = topic_it->second;
+    if (revision < cur_state.revision) {
+        return stm_update_error{fmt::format(
+          "topic {} rev {}: lifecycle update with obsolete revision {}",
+          topic,
+          cur_state.revision,
+          revision)};
+    } else if (revision > cur_state.revision) {
+        if (
+          cur_state.lifecycle_state != topic_state::lifecycle_state_t::purged) {
+            return stm_update_error{fmt::format(
+              "topic {} rev {} not fully purged, but already trying to "
+              "register revision {}",
+              topic,
+              cur_state.revision,
+              revision)};
+        }
+
```

Comment:
The allowing of newer revisions in some cases is making it really difficult to validate correctness IMO. Is there a good reason for us to not just reject requests of different revision? If not, i feel very strongly about rejecting different revisions

If there is a good reason, i'd appreciate comments about a mental model for this that makes it easier to reason about.

Reply:
This is precisely the transition that allows us to switch to a newer revision. This has to happen somewhere, right?